rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py

```diff
@@ -0,0 +1,671 @@
+        batch_length_T = batch_length_T or self.batch_length_T
+
+        # Sample for each module independently.
+        if replay_mode == "independent":
+            return self._sample_independent(
+                batch_size_B=batch_size_B,
+                batch_length_T=batch_length_T,
+                n_step=n_step,
+                gamma=gamma,
+                include_infos=include_infos,
+                include_extra_model_outputs=include_extra_model_outputs,
+                modules_to_sample=modules_to_sample,
+            )
+        else:
+            return self._sample_synchonized(
+                batch_size_B=batch_size_B,
+                batch_length_T=batch_length_T,
+                n_step=n_step,
+                gamma=gamma,
+                include_infos=include_infos,
+                include_extra_model_outputs=include_extra_model_outputs,
+                modules_to_sample=modules_to_sample,
+            )
+
+    def _sample_independent(
+        self,
+        batch_size_B: Optional[int],
+        batch_length_T: Optional[int],
+        n_step: Optional[Union[int, Tuple]],
+        gamma: float,
+        include_infos: bool,
+        include_extra_model_outputs: bool,
+        modules_to_sample: Optional[List[Union[str, int]]],
+    ) -> SampleBatchType:
+
+        # Sample the n-step if necessary.
+        if isinstance(n_step, tuple):
+            # Use random n-step sampling.
+            random_n_step = True
+        else:
+            actual_n_step = n_step or 1
+            random_n_step = False
+
+        ret = {}
+        # TODO (simon): Ensure that the module has data and if not, skip it.
+        #  TODO (sven): Should we then error out or skip? I think the Learner
+        #  should handle this case when a module has no train data.
+        for module_id in modules_to_sample or self._module_indices.keys():
+            # Rows to return.
+            observations = [[] for _ in range(batch_size_B)]
+            next_observations = [[] for _ in range(batch_size_B)]
+            actions = [[] for _ in range(batch_size_B)]
+            rewards = [[] for _ in range(batch_size_B)]
+            is_terminated = [False for _ in range(batch_size_B)]
+            is_truncated = [False for _ in range(batch_size_B)]
+            weights = [1.0 for _ in range(batch_size_B)]
+            n_steps = [[] for _ in range(batch_size_B)]
+            # If `info` should be included, construct also a container for them.
+            if include_infos:
+                infos = [[] for _ in range(batch_size_B)]
+            # If `extra_model_outputs` should be included, construct a container for
+            # them.
+            if include_extra_model_outputs:
+                extra_model_outputs = [[] for _ in range(batch_size_B)]
+            B = 0
+            while B < batch_size_B:
+                # Now sample from the single-agent timesteps.
+                index_tuple = self._module_indices[module_id][
```

Comment:
Let's explain here again what an "index" is: A 4-tuple/"quad"(?) (<- let's always use 4-tuple for clarity) of ... ?